File: python/engine/retry.py
from __future__ import annotations

import math
from typing import Optional

from .models import RetryPolicy


def compute_backoff_seconds(policy: RetryPolicy, attempt: int) -> int:
    """
    Compute exponential backoff with a maximum cap.

    attempt is 1-based (first retry attempt is attempt=1).
    """

    if attempt <= 0:
        return policy.initial_interval_seconds
    interval = policy.initial_interval_seconds * (policy.backoff_coefficient ** (attempt - 1))
    interval = min(interval, policy.max_interval_seconds)
    return int(math.ceil(interval))

File: go/engine/retry.go
package engine

import "math"

// ComputeBackoffSeconds calculates exponential backoff duration for attempts.
// attempt is 1-based.
func ComputeBackoffSeconds(policy RetryPolicy, attempt int) int {
	if attempt <= 0 {
		return policy.InitialIntervalSeconds
	}
	interval := float64(policy.InitialIntervalSeconds) * math.Pow(policy.BackoffCoefficient, float64(attempt-1))
	if interval > float64(policy.MaxIntervalSeconds) {
		interval = float64(policy.MaxIntervalSeconds)
	}
	return int(math.Ceil(interval))
}
